src/main/scala/dsptools/Chisel3TestersCompatibility.scala

```diff
@@ -0,0 +1,14 @@
+package chisel3.iotesters
```

Comment:
Yeah, software engineering quality is a continuing work in progress. Hopefully things will get better.

flatten: yes, the implementation is correct, but ultimately it probably shouldn't be a member on Data. The alternative isn't necessarily to reimplement it, but to refactor or restructure use sites to use cleaner, more local methods, like associating directionality with Aggregates rather than inspecting the leaf elements. For example, PeekPokeTester could recursively pack / unpack bits into a Bundle, instead of flattening a blackbox Data to leaf Elements. From a structural standpoint, this would expose a packing / unpacking function that can have arbitrary implementation for each data type, rather than confining every data structure to be some collection of Elements.

As an example: what happens if you flatten something that includes an Analog type? Or how would you flatten an union type?

Hiding API: if it's part of the public API, someone is going to use it, and it makes it difficult to remove / change later if we discover it's bad. The closest thing we're doing is the experimental package, but ultimately every API item comes with an expectation of continued support. And people complain when things change, whether it was intended as a "public" API or not...

Obviously, the API we've exposed isn't perfect, but it would also be nice to work with us to make a better, more featured API in a principled way rather than punching holes into abstraction boundaries. There's usually a reason we've defined the abstraction boundaries in the way we did, and in a way that might not be immediately obvious to someone who isn't developing infrastructure.

For the PeekPokeTester API: I guess it could be structured better? But chisel-testers doesn't seem to see as much debate compared to the rest of core chisel. Much of the software engineering progress on core chisel has been because of the development effort by many, many people with different backgrounds and opinions.

addPin: you're right in that we should have had a replacement available when it was removed (I think this is what Record is for? Also keep in mind that the core Chisel team is limited so we can't implement everything everyone wants, but you're welcome to contribute =) ). Superusers and leaky abstractions also presents a problem: they will do something that pushes the boundaries of the system in a way the developers never intended, relying on internal implementation details such that continued development becomes impossible without changing externally visible behavior. Or they do things in a complex or clever way that ends up being as structurally sound as a house of cards. Ultimately there's a compromise between how fast internal development can happen, how many features are exposed, and how correct everything is. Chisel2 probably was pretty much in the "gradware" region of that compromise (expose everything, move fast and break things) while Chisel3 aims to be more stable and principled. However, that means not exposing everything like was done in Chisel2, which unfortunately appears to users as removal of features.